surveyor/conn_pool.go

```diff
@@ -0,0 +1,254 @@
+		return
+	}
+	pc.cp.Unlock()
+}
+
+type natsConnPool struct {
+	sync.Mutex
+	cache        map[string]*pooledNatsConn
+	logger       *logrus.Logger
+	group        *singleflight.Group
+	natsDefaults *natsContextDefaults
+	natsOpts     []nats.Option
+}
+
+func newNatsConnPool(logger *logrus.Logger, natsDefaults *natsContextDefaults, natsOpts []nats.Option) *natsConnPool {
+	return &natsConnPool{
+		cache:        map[string]*pooledNatsConn{},
+		group:        &singleflight.Group{},
+		logger:       logger,
+		natsDefaults: natsDefaults,
+		natsOpts:     natsOpts,
+	}
+}
+
+func (cp *natsConnPool) Get(cfg *natsContext) (*pooledNatsConn, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("nats context must not be nil")
+	}
+
+	// copy cfg
+	cfg = cfg.copy()
+
+	// set defaults
+	if cfg.Name == "" {
+		cfg.Name = cp.natsDefaults.Name
+	}
+	if cfg.URL == "" {
+		cfg.URL = cp.natsDefaults.URL
+	}
+	if cfg.TLSCA == "" {
+		cfg.TLSCA = cp.natsDefaults.TLSCA
+	}
+	if cfg.TLSCert == "" {
+		cfg.TLSCert = cp.natsDefaults.TLSCert
+	}
+	if cfg.TLSKey == "" {
+		cfg.TLSKey = cp.natsDefaults.TLSKey
+	}
+
+	// get hash
+	key, err := cfg.hash()
+	if err != nil {
+		return nil, err
+	}
+
+GetFromPool:
+	conn, err, _ := cp.group.Do(key, func() (interface{}, error) {
+		cp.Lock()
+		pooledConn, ok := cp.cache[key]
+		if ok && pooledConn.nc.IsConnected() {
+			cp.Unlock()
+			return pooledConn, nil
+		}
+		cp.Unlock()
+
+		opts := cp.natsOpts
+		opts = append(opts, func(options *nats.Options) error {
+			if cfg.Name != "" {
+				options.Name = cfg.Name
+			}
+			if cfg.Token != "" {
+				options.Token = cfg.Token
+			}
+			if cfg.Username != "" {
+				options.User = cfg.Username
+			}
+			if cfg.Password != "" {
+				options.Password = cfg.Password
+			}
+			return nil
+		})
+
+		if cfg.JWT != "" && cfg.Seed != "" {
+			opts = append(opts, nats.UserJWTAndSeed(cfg.JWT, cfg.Seed))
+		}
+
+		if cfg.Nkey != "" {
+			opt, err := nats.NkeyOptionFromSeed(cfg.Nkey)
+			if err != nil {
+				return nil, fmt.Errorf("unable to load nkey: %v", err)
+			}
+			opts = append(opts, opt)
+		}
+
+		if cfg.Credentials != "" {
+			opts = append(opts, nats.UserCredentials(cfg.Credentials))
+		}
+
+		if cfg.TLSCA != "" {
+			opts = append(opts, nats.RootCAs(cfg.TLSCA))
+		}
+
+		if cfg.TLSCert != "" && cfg.TLSKey != "" {
+			opts = append(opts, nats.ClientCert(cfg.TLSCert, cfg.TLSKey))
+		}
+
+		nc, err := nats.Connect(cfg.URL, opts...)
+		if err != nil {
+			return nil, err
+		}
+		cp.logger.Infof("%s connected to NATS Deployment: %s", cfg.Name, nc.ConnectedAddr())
+
+		connection := &pooledNatsConn{
+			nc:  nc,
+			cp:  cp,
+			key: key,
+		}
+
+		cp.Lock()
+		cp.cache[key] = connection
+		cp.Unlock()
+
+		return connection, err
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	connection, ok := conn.(*pooledNatsConn)
+	if !ok {
+		return nil, fmt.Errorf("not a pooledNatsConn")
+	}
+
+	cp.Lock()
+	if connection.closed {
+		// ReturnToPool closed this while lock not held, try again
```

Comment:
this is the `pooledNatsConn` - it would only get closed here if one pooled connection was already leased out, and was returned in the very brief period between pulling it from the cache and here.

Reply:
wouldn't hurt to add a max bound though - could guard against a future bug

Reply:
added max bound of 10

Reply:
Yeah, I know this should not happen, but guarding against future bugs is what I was thinking about :)